### PR TITLE
Add actual paid date and external identifier to data available to invoice templates

### PR DESCRIPTION
**Pull request by jamie-tillman**
Overview
----------------------------------------
I made this edit on a local customization and figured it might be useful to others.  This checks to see if the invoice is paid and, if so, looks up the date of the last financial transactions.  It also provides the contact's external identifier.

Before
----------------------------------------
These values were not provided to the templates.  No way to show the actual paid date or the external identifier on the invoice.

After
----------------------------------------
Basic lookup of financial transaction data if paid amount is equal to the invoice amount.  Adds the external identifier property to the other data values being stored, using the same object as the 'display_name' value.

### CODE REVIEW

**Comment by MegaphoneJon** [comment]
@jamie-tillman This makes a lot of sense - and is a far sight better than my technique of making a bunch of Smarty API calls at the top of the template.

I'll tell you right now that Eileen isn't going to let any template variables changes through without a test though, since this area of the code is, well, not good.

**Comment by jamie-tillman** [comment]
> I'll tell you right now that Eileen isn't going to let any template variables changes through without a test though, since this area of the code is, well, not good.

Fair enough.  I'll try to find some time to prepare a unit test.  I'm not up to speed on how to do that anyway, so it's a good learning opportunity.

**Comment by mattwire** [comment]
@jamie-tillman Given that `contactId` is being assigned to the template we should be able to get `external_identifier` as a contact token and not rely on `tplParams` which is a bit more random/difficult to support.

For the invoice_paid_date, is that part of the contribution entity? In which case it would be nice to do that via a contribution token as well.

**Comment by jamie-tillman** [comment]
> @jamie-tillman Given that `contactId` is being assigned to the template we should be able to get `external_identifier` as a contact token and not rely on `tplParams` which is a bit more random/difficult to support.

You're right.  And I just tested and can confirm that.  This code tweak is fairly old, and somehow that option got missed when we made the adjustment.  Yay for less code!
 
> For the invoice_paid_date, is that part of the contribution entity? In which case it would be nice to do that via a contribution token as well.

As I recall, the "paid" date is not part of the contribution's data (at least not what was retrievable via the api), which is why we added it as a hack.  That said, I do know there are a "receipt_date" and a "received_date", and IIRC neither is reliable for obtaining the date the invoice was actually paid.  This may have changed, as our hack was made several versions ago and has been manually ported forward as we upgraded.

That manual porting activity is the main driver behind my PR.  I was hoping to save myself some aggravation by getting it into the main codebase and contribute to the community at the same time.

**Comment by mattwire** [comment]
@jamie-tillman Ok, so can we drop `external_identifier` from this PR as better to use already supported token method.

For the invoicePaidDate it seems pretty unique to this invoice code so can we get a comment around the block you've added to get the date explaining why/how. There are tests in `CRM_Contribute_Form_Task_InvoiceTest` which you could use as a base to test this new parameter.

**Comment by MegaphoneJon** [comment]
@jamie-tillman Not to discourage you from contributing - it's appreciated! - but I wanted to show my technique for handling this without touching code.

Here's a snippet from a customized invoice I use that shows the technique of using the Smarty `crmAPI` extension with the built-in `capture` function.  The variables named immediate after `capture` are available for use in subsequent lines of the template.

```
{crmAPI var='clookup' entity='Contribution' action='getsingle' return="contact_id,total_amount,custom_228,custom_244,custom_245,custom_246,custom_247,custom_249,custom_252,custom_250" id=$id}
{capture "contact_id" assign=contact_id}{$clookup.contact_id}{/capture}
{capture "billing_period" assign=billing_period}{$clookup.custom_228}{/capture}
{crmAPI var='ilookup' entity='Invoicegen' action='invoicedata' billing_period=$billing_period contact_id=$contact_id}
{crmAPI var='org' entity='Contact' action='getsingle' return="custom_156,custom_236,custom_206,country" id=$contact_id}
{crmAPI var='licensee_cid' entity='Relationship' action='getvalue' return="contact_id_a" contact_id_b=$contact_id}
{crmAPI var='licensee_info' entity='Contact' action='getsingle' return="display_name,email" id=$licensee_cid}
{capture "tb_url" assign=tb_url}{$org.custom_156}{/capture}
{capture "number_of_members" assign=number_of_members}{$clookup.custom_249}{/capture}
{capture "biannual_fee" assign=biannual_fee}{$clookup.custom_250}{/capture}
```

**Comment by MegaphoneJon** [comment]
@jamie-tillman It IS fundamental - see https://lab.civicrm.org/dev/financial/-/issues/48, which leads back to one of the toughest nuts yet to be cracked, https://lab.civicrm.org/dev/financial/-/issues/43.

tl;dr: When Civi was written, there was no double-entry accounting.  No payments, line items, financial transactions - nothing but Contribution records.  Introduction of accounting principles in Civi 4.3 was a much-needed feature, but efforts were made not to disrupt the workflow of organizations who didn't need it.

These days, there's widespread agreement that we need to make "payments" a first-class entity in Civi - but there are many fields still on Contribution that belong on payments (receive_date, check_number, etc.) and it's a lot of work to preserve existing workflows while removing those fields.

**Comment by eileenmcnaughton** [comment]
@jamie-tillman I checked and ``{contact.external_identifier} `` DOES work - see https://github.com/civicrm/civicrm-core/pull/20867

For the paidDate  - I think it would be best to add it in an extension for now. To get it into core we would need it to be handled, and unit tested, for BOTH of the ways in which we handle tokens (not via smarty). Unfortunately we have a legacy token processor and a less legacy token processor in play at the moment which we are working to get in sync (we need to get to the point where we have unit tests proving they both present exactly the same parameters). 

I'm going to close this for now - if you want to bring it back then ANYTHING in the token space would not be reviewable without being able to demonstrate test cover
